### PR TITLE
feat: adding replaceServerVariablesWithValues function

### PR DESCRIPTION
**Pull request by amrutprabhu**
**Description**
This is part of the pull request for adding Kafka broker settings. 
the `replaceServerVariablesWithValues` function is currently located in the nodejs-templae, but is also required in the spring-cloud-stream-template. 

@derberg had recommended moving this function to `generator-filters` as a common place for this common function. 

**Related issue(s)**
Helps to solve this PR: https://github.com/asyncapi/java-spring-cloud-stream-template/pull/125

### CODE REVIEW

File: src/customFilters.js [hunk 1]
```diff
@@ -174,3 +174,52 @@ function docline(field, fieldName, scopePropName) {
...
 filter.docline = docline;
+
+/*
+*  Helper function to replace, server variables in the url with actual values
+*  @url {string} - url string
+*  @serverserverVariables {Object} - Varibles model map
+*  @returns {string} 
+*
+*/
+
```
Comment (magicmatatjahu) [inline src/customFilters.js:185]:
```suggestion
```

**Review by magicmatatjahu — COMMENTED** [review]
@amrutprabhu Thanks for contribution! Could you also add the tests for new function in the `customFilters.test.js` file?

**Comment by derberg** [comment]
@amrutprabhu lemme know if you need some help with tests, and also have a look at failing linter.

[here](https://github.com/asyncapi/generator-filters/blob/master/src/customFilters.test.js#L50) you can find an example of test with message model object, so you know how to do a test where [variable model](https://github.com/asyncapi/parser-js/blob/master/lib/models/server-variable.js) is involved

**Comment by amrutprabhu** [comment]
@derberg I am new to javascript testing.. But I will try ..

**Comment by amrutprabhu** [comment]
@derberg @magicmatatjahu Can you have a look at it.. ?

**Review by derberg — COMMENTED** [review]
It looks good, but improve tests as we should be consistent in there:

do this `const ServerVariable = require('@asyncapi/parser/lib/models/server-variable');`

then, instead of for example `const serverVariables = JSON.parse('{"port":{"_json":{"enum":["8080","8883"]}}}'); `

you should do `const serverVariables = new ServerVariable( { port: { enum: ["8080","8883"] } }); `

File: src/customFilters.test.js [hunk 2]
```diff
@@ -189,4 +190,24 @@ test('.getHeadersExamples() should return examples for headers - case when at le
       example: { bar: 'foo' },
     },
   ]);
+});
+
+test('.replaceServerVariablesWithValues() should replace placeholder with default value', t => {
+  const is = t.is;
+  const inputUrl = 'localhost:{port}'; 
+  const serverVariables = new ServerVariable( '{"port":{"default":"9092","enum":["8080","8883"]}}');
```
Comment (magicmatatjahu) [inline src/customFilters.test.js:198]:
```suggestion
  const serverVariables = new ServerVariable({"port":{"default":"9092","enum":["8080","8883"]}}); 
```

File: src/customFilters.test.js [hunk 3]
```diff
@@ -189,4 +190,24 @@ test('.getHeadersExamples() should return examples for headers - case when at le
...
+  const expected = 'localhost:9092';
+  const value    = replaceServerVariablesWithValues(inputUrl, serverVariables)
+  
+  is(value, expected);
+});
+
+test('.replaceServerVariablesWithValues() should replace placeholder with first enum value when no default is specified', t => {
+  const is = t.is;
+  const inputUrl = 'localhost:{port}'; 
+  const serverVariables = new ServerVariable( '{"port":{"enum":["8080","8883"]}}');
```
Comment (magicmatatjahu) [inline src/customFilters.test.js:208]:
```suggestion
  const serverVariables = new ServerVariable({"port":{"enum":["8080","8883"]}}); 
```

**Review by magicmatatjahu — CHANGES_REQUESTED** [review]
Tests fail. You should put JSON object (not stringified version) to the `ServerVariable` class.

**Comment by amrutprabhu** [comment]
@derberg Apparently, the test fails after using the servervariables. How can I solve this? 
sorry for my lack of javascript knowledge.

**Comment by amrutprabhu** [comment]
> @derberg Apparently, the test fails after using the servervariables. How can I solve this?
> sorry for my lack of javascript knowledge.

So I did a little debugging. Just printed the jsono format of the servervaraibels. And I get the following. 
```
{"_json":{"port":{"default":"9092","enum":["8080","8883"]}}}
```
Which apparently does not match what the function was expecting from https://github.com/asyncapi/nodejs-template/blob/fff91aa1f76db4682332f16bb91cdd78fb536ceb/template/config/common.yml#L41

File: src/customFilters.test.js [hunk 4]
```diff
@@ -189,4 +190,24 @@ test('.getHeadersExamples() should return examples for headers - case when at le
       example: { bar: 'foo' },
     },
   ]);
+});
+
+test('.replaceServerVariablesWithValues() should replace placeholder with default value', t => {
+  const is = t.is;
+  const inputUrl = 'localhost:{port}'; 
+  const serverVariables = new ServerVariable({port: { default:"9092", enum: ["8080","8883"]}});
```
Comment (magicmatatjahu) [inline src/customFilters.test.js:198]:
```suggestion
  const serverVariables = { port: new ServerVariable({ default:"9092", enum: ["8080","8883"] }) }; 
```

File: src/customFilters.test.js [hunk 5]
```diff
@@ -189,4 +190,24 @@ test('.getHeadersExamples() should return examples for headers - case when at le
...
+  const expected = 'localhost:9092';
+  const value    = replaceServerVariablesWithValues(inputUrl, serverVariables)
+  
+  is(value, expected);
+});
+
+test('.replaceServerVariablesWithValues() should replace placeholder with first enum value when no default is specified', t => {
+  const is = t.is;
+  const inputUrl = 'localhost:{port}'; 
+  const serverVariables = new ServerVariable({ port: { enum: ["8080","8883"] } });
```
Comment (magicmatatjahu) [inline src/customFilters.test.js:208]:
```suggestion
  const serverVariables = { port: new ServerVariable({ enum: ["8080","8883"] }) }; 
```

**Review by magicmatatjahu — COMMENTED** [review]
Sorry 😅  You should pass variables as map/dictionary, not instance of ServerVariable class.

File: src/customFilters.js [hunk 6]
```diff
@@ -183,8 +183,8 @@ filter.docline = docline;
 */
 function replaceServerVariablesWithValues(url, serverVariables) {
   const getVariablesNamesFromUrl = (inputUrl) => {
-    let result = [],
-      array;
+    const result = [], array;
```
Comment (magicmatatjahu) [inline src/customFilters.js:186]: `array` should be initialised as `let`.

**Review by magicmatatjahu — APPROVED** [review]
@amrutprabhu 👏🏼  Thanks for contribution!

**Comment by amrutprabhu** [comment]
pheww.. Those were some long list of problems. :smile: 
Can you have a look now @derberg and @magicmatatjahu

**Comment by amrutprabhu** [comment]
hey @magicmatatjahu,
I see that the docs that I had updated are not reflecting. 
It looks like the asyncapi-bot overrode my changes.

I see my change In the commit prior to it here: https://github.com/asyncapi/generator-filters/blob/f171ad4c3172f9446de26a2b80d2ffa0a40b4ff0/docs/api.md

**Comment by magicmatatjahu** [comment]
@amrutprabhu Thanks for info! I will look at it, because it's very strange 🤔  PR with new docs should help :)

**Comment by derberg** [comment]
@amrutprabhu I just noticed you updated API manually and not with `npm run docs`. @magicmatatjahu you looking into this or shall I?

**Comment by magicmatatjahu** [comment]
I will, but I wanna understand why PR with bumped version removed the docs, if in pipeline we have appropriate script https://github.com/asyncapi/generator-filters/blob/c42fbe9f75860e055e9de4f80d892ec87ed26611/.github/workflows/if-nodejs-version-bump.yml#L31 🤔

**Comment by derberg** [comment]
> but I wanna understand why PR with bumped version removed the docs

this is exactly why there were removed, cause `npm run docs` ran on version bump workflow. So the problem is somewhere with JSdoc imho, there must be some error that makes jsdoc2md to ignore it, thus remove if was added manually

**Comment by magicmatatjahu** [comment]
PR -> https://github.com/asyncapi/generator-filters/pull/49

**Pull request merged**